metadata-ingestion/src/datahub/ingestion/source/sql/clickhouse.py

```diff
@@ -0,0 +1,631 @@
+import json
+import textwrap
+from collections import defaultdict
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple, Union
+
+import clickhouse_driver  # noqa: F401
+import clickhouse_sqlalchemy.types as custom_types
+from clickhouse_sqlalchemy.drivers import base
+from clickhouse_sqlalchemy.drivers.base import ClickHouseDialect
+from sqlalchemy import create_engine, text
+from sqlalchemy.engine import reflection
+from sqlalchemy.sql import sqltypes
+from sqlalchemy.types import BOOLEAN, DATE, DATETIME, INTEGER
+
+import datahub.emitter.mce_builder as builder
+from datahub.configuration.time_window_config import BaseTimeWindowConfig
+from datahub.emitter import mce_builder
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.workunit import MetadataWorkUnit
+from datahub.ingestion.source.sql.sql_common import (
+    BasicSQLAlchemyConfig,
+    SQLAlchemySource,
+    SqlWorkUnit,
+    logger,
+    register_custom_type,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.dataset import UpstreamLineage
+from datahub.metadata.com.linkedin.pegasus2avro.metadata.snapshot import DatasetSnapshot
+from datahub.metadata.com.linkedin.pegasus2avro.mxe import MetadataChangeEvent
+from datahub.metadata.com.linkedin.pegasus2avro.schema import (
+    ArrayTypeClass,
+    MapTypeClass,
+    NumberTypeClass,
+    StringTypeClass,
+    UnionTypeClass,
+)
+from datahub.metadata.schema_classes import (
+    ChangeTypeClass,
+    DatasetLineageTypeClass,
+    DatasetPropertiesClass,
+    DatasetSnapshotClass,
+    UpstreamClass,
+)
+
+# adding extra types not handled by clickhouse-sqlalchemy 0.1.7
+base.ischema_names["DateTime64(0)"] = DATETIME
+base.ischema_names["DateTime64(1)"] = DATETIME
+base.ischema_names["DateTime64(2)"] = DATETIME
+base.ischema_names["DateTime64(3)"] = DATETIME
+base.ischema_names["DateTime64(4)"] = DATETIME
+base.ischema_names["DateTime64(5)"] = DATETIME
+base.ischema_names["DateTime64(6)"] = DATETIME
+base.ischema_names["DateTime64(7)"] = DATETIME
+base.ischema_names["DateTime64(8)"] = DATETIME
+base.ischema_names["DateTime64(9)"] = DATETIME
+base.ischema_names["Date32"] = DATE
+base.ischema_names["Bool"] = BOOLEAN
+base.ischema_names["Nothing"] = sqltypes.NullType
+base.ischema_names["Int128"] = INTEGER
+base.ischema_names["Int256"] = INTEGER
+base.ischema_names["UInt128"] = INTEGER
+base.ischema_names["UInt256"] = INTEGER
+
+register_custom_type(custom_types.common.Array, ArrayTypeClass)
+register_custom_type(custom_types.ip.IPv4, NumberTypeClass)
```

Comment:
IPv6 is mapped as string but this one as Number, why?

Reply:
In ClickHouse those two IP types are based on different data type.
More info in the documentation [IPv4](https://clickhouse.com/docs/en/sql-reference/data-types/domains/ipv4/) and [IPv6](https://clickhouse.com/docs/en/sql-reference/data-types/domains/ipv6/).

Reply:
thanks for the clarificaiton